checker/checker.go

```diff
@@ -75,113 +84,127 @@ func (c *checker) Check(mod *parser.Module) error {
 				}
 			}
 
-			// Binds may be declared inside the body of a function, and their target
-			// identifiers will also collide with any other declarations, so we must
-			// check if there are any duplicate declarations there.
-			parser.Inspect(fun, func(node parser.Node) bool {
-				n, ok := node.(parser.Block)
-				if !ok {
-					return true
-				}
-
-				for _, stmt := range n.List() {
-					call := stmt.Call
-					binds := call.Binds
-					if binds == nil {
-						continue
-					}
-
-					// mount scratch "/" as default
-					if binds.Ident != nil {
-						c.registerDecl(mod.Scope, binds.Ident, binds)
-					}
-
-					// mount scratch "/" as (target default)
-					if binds.List != nil {
-						for _, b := range binds.List.List {
-							c.registerDecl(mod.Scope, b.Target, binds)
-						}
-					}
-
-					// Link to its lexical scope.
-					binds.Lexical = fun
-
-					err := c.bindEffects(mod.Scope, n.ObjType(), call)
-					if err != nil {
-						c.errs = append(c.errs, err)
-					}
-				}
-
-				return true
-			})
-
-			// Do not walk the function node's children since we already walked.
-			return false
-		}
+			// Propagate scope and type into its BlockStmt.
+			if fun.Body != nil {
+				fun.Body.Scope = fun.Scope
+				fun.Body.ObjType = fun.Type.ObjType
+			}
+		},
+		// ImportDecl's BlockStmts have module-level scope and inherits the function
+		// literal type.
+		func(_ *parser.ImportDecl, lit *parser.FuncLit) {
+			lit.Body.Scope = mod.Scope
+			lit.Body.ObjType = lit.Type.ObjType
+		},
+		// FuncDecl's BlockStmts have function-level scope and inherits the function
+		// literal's type.
+		func(fun *parser.FuncDecl, lit *parser.FuncLit) {
+			if lit.Type.ObjType == parser.Option {
+				return
+			}
+			lit.Body.Scope = fun.Scope
+			lit.Body.ObjType = lit.Type.ObjType
+		},
+		// Function literals propagate its scope to its children.
+		func(parentLit *parser.FuncLit, lit *parser.FuncLit) {
+			lit.Body.Scope = parentLit.Body.Scope
+			lit.Body.ObjType = parentLit.Body.ObjType
+		},
+		// WithOpt's function literals need to infer its secondary type from its
+		// parent call statement. For example, `run with option { ... }` has a
+		// `option` type function literal, but infers its type as `option::run`.
+		func(fun *parser.FuncDecl, call *parser.CallStmt, with *parser.WithOpt, lit *parser.FuncLit) {
```

Comment:
> First, doesn't with also need to supply the function register to non-literal options, eg with readonly?
Technically yes, but I don't think any non-literal options uses function registers atm. You're also not allowed to define binds when using a non-literal builtin in this short-hand style. Though I believe we can "fulfill" this by doing what you suggest later, which is to also have a `Lexical` pointer in `WithOpt`.

> the FuncDecl and FuncLit are sharing scope
The conceptual level is that the `FuncLit` should have a sub-scope deriving from `FuncDecl`'s scope. Implementation-wise, you can't define new decls within `FuncLit` atm, so there isn't any need to create a sub-scope. Thinking about this more, I think FuncLits shares scope / derives sub-scope, but they each have their own registers (except option func lits). The only exception is `with` that negotiates access between option func lits and the parent func's registers.

> would it make sense to memoize type information and lexical scope onto the WithOpt itself, so that later Matchers can use that to find the Lexical func?
I think this makes great sense. I need to find a way to pass down either `*FuncDecl` or `*WithOpt` to able to access the `Lexical` pointer though.

Reply:
FYI, renamed `Lexical` to `Closure` following the conversation on slack.